consensus/spos/bls/v2/subroundEndRound.go

```diff
@@ -0,0 +1,955 @@
+}
+
+func (sr *subroundEndRound) prepareBroadcastBlockData() error {
+	miniBlocks, transactions, err := sr.BlockProcessor().MarshalizedDataToBroadcast(sr.GetHeader(), sr.GetBody())
+	if err != nil {
+		return err
+	}
+
+	getEquivalentProofSender := sr.getEquivalentProofSender()
+	go sr.BroadcastMessenger().PrepareBroadcastBlockDataWithEquivalentProofs(sr.GetHeader(), miniBlocks, transactions, []byte(getEquivalentProofSender))
+
+	return nil
+}
+
+func (sr *subroundEndRound) waitForProof() bool {
+	shardID := sr.ShardCoordinator().SelfId()
+	headerHash := sr.GetData()
+	if sr.EquivalentProofsPool().HasProof(shardID, headerHash) {
+		return true
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), sr.RoundHandler().TimeDuration())
+	defer cancel()
+
+	for {
+		select {
+		case <-time.After(time.Millisecond):
+			if sr.EquivalentProofsPool().HasProof(shardID, headerHash) {
+				return true
+			}
+		case <-ctx.Done():
+			return false
+		}
+	}
+}
+
+func (sr *subroundEndRound) finalizeConfirmedBlock() bool {
+	if !sr.waitForProof() {
+		return false
+	}
+
+	ok := sr.ScheduledProcessor().IsProcessedOKWithTimeout()
+	// placeholder for subroundEndRound.doEndRoundJobByLeader script
+	if !ok {
+		return false
+	}
+
+	err := sr.commitBlock()
+	if err != nil {
+		return false
+	}
+
+	sr.SetStatus(sr.Current(), spos.SsFinished)
+
+	sr.worker.DisplayStatistics()
+
+	log.Debug("step 3: Body and Header have been committed")
+
+	msg := fmt.Sprintf("Added proposed block with nonce  %d  in blockchain", sr.GetHeader().GetNonce())
+	log.Debug(display.Headline(msg, sr.SyncTimer().FormattedCurrentTime(), "+"))
+
+	sr.updateMetricsForLeader()
+
+	return true
+}
+
+func (sr *subroundEndRound) sendProof() {
+	if !sr.shouldSendProof() {
+		return
+	}
+
+	bitmap := sr.GenerateBitmap(bls.SrSignature)
+	err := sr.checkSignaturesValidity(bitmap)
+	if err != nil {
+		log.Debug("sendProof.checkSignaturesValidity", "error", err.Error())
+		return
+	}
+
+	// Aggregate signatures, handle invalid signers and send final info if needed
+	bitmap, sig, err := sr.aggregateSigsAndHandleInvalidSigners(bitmap)
+	if err != nil {
+		log.Debug("sendProof.aggregateSigsAndHandleInvalidSigners", "error", err.Error())
+		return
+	}
+
+	roundHandler := sr.RoundHandler()
+	if roundHandler.RemainingTime(roundHandler.TimeStamp(), roundHandler.TimeDuration()) < 0 {
+		log.Debug("sendProof: time is out -> cancel broadcasting final info and header",
+			"round time stamp", roundHandler.TimeStamp(),
+			"current time", time.Now())
+		return
+	}
+
+	// broadcast header proof
+	err = sr.createAndBroadcastProof(sig, bitmap)
+	if err != nil {
+		log.Warn("sendProof.createAndBroadcastProof", "error", err.Error())
+	}
+}
+
+func (sr *subroundEndRound) shouldSendProof() bool {
+	if sr.EquivalentProofsPool().HasProof(sr.ShardCoordinator().SelfId(), sr.GetData()) {
+		log.Debug("shouldSendProof: equivalent message already processed")
+		return false
+	}
+
+	return sr.IsSelfInConsensusGroup()
```

Comment:
Does this mean that at block creation like every validator is sending the proofs ? 

Couldn't we wait a little bit of time, before sending?

Reply:
This is the equivalent proof mechanism where we optimize the propagation of equivalent proofs.
Only the validators that have not yet received a proof, will send it, and there is a protection so that the messages do not flood the system (can be seen in the performance results that there is no noticeable increase in outgoing/incoming network used bandwidth)